File: src/main/go/leetcode/lists/hot100/LeetCode20.go
package hot100

func isValid(s string) bool {
	stack := make([]byte, 0, len(s))
	for i := 0; i < len(s); i++ {
		c := s[i]
		if c == '(' || c == '{' || c == '[' {
			stack = append(stack, c)
		} else {
			if len(stack) == 0 {
				return false
			}
			last := stack[len(stack)-1]
			if c == ')' && last == '(' ||
				c == ']' && last == '[' ||
				c == '}' && last == '{' {
				stack = stack[:len(stack)-1]
			} else {
				return false
			}
		}
	}
	return len(stack) == 0
}

File: java/src/leetcode/problems/LeetCode20.java
package leetcode.problems;

import java.util.Map;
import java.util.Stack;

/**
 * @author wujingxinit@outlook.com
 * @date 2024/4/9 15:22
 */
public class LeetCode20 {

    public boolean isValid(String s) {
        if (s.length() % 2 != 0) return false;
        Stack<Character> st = new Stack<>();
        Map<Character, Character> map = Map.of(')', '(', ']', '[', '}', '{');
        for (char c : s.toCharArray()) {
            if (map.containsKey(c)) {
                if (st.isEmpty() || st.pop() != map.get(c)) return false;
            } else st.push(c);
        }
        return st.isEmpty();
    }

    class Solution_OPT {

        public boolean isValid(String s) {
            if (s.length() % 2 != 0) return false;
            char[] stack = new char[s.length()];
            int top = 0;
            char[] chs = s.toCharArray();
            for (char c : chs) {
                if (c == '(') {
                    stack[top++] = ')';
                } else if (c == '[') {
                    stack[top++] = ']';
                } else if (c == '{') {
                    stack[top++] = '}';
                } else {
                    if (top == 0 || stack[--top] != c) {
                        return false;
                    }
                }
            }
            return top == 0;
        }
    }
}
